SMF Go generator assumes flatbuffer names and directory structures

**Describe the bug**

The go code generator within smfc is a nice feature, as it allows for some quick prototyping and use of non-Linux platforms for testing.

That being said, the code generator has some hardcoded paths in it that cause `go build` to reject the output code.  Example failed `go build` output includes:

```
# not_a_demo
../not_a_demo/not_demo_service.smf.fb.go:13:24: undefined: smf.Request
../not_a_demo/not_demo_service.smf.fb.go:37:32: undefined: demo
../not_a_demo/not_demo_service.smf.fb.go:48:68: undefined: not_a_demo
../not_a_demo/not_demo_service.smf.fb.go:53:9: undefined: smf.GetRootAsResponse
```

The first two and the last one appear to be the result of hardcoding for specific expected types that don't always exist (`Request`, `Response`, `Demo`). 

The third one uses the same package as the current flatbuffer protocol file to identify a the response flatbuffer type, which is fine if the generated service is in a different package.  In my example, below, this is not the case.

**To Reproduce**
1. Set `GOPATH` to something appropriate, and `mkdir -p $GOPATH/src`
2. `cd $GOPATH/src`, and get dependencies:

```
cd $GOPATH/src
go get github.com/google/flatbuffers/go
go get github.com/smfrpc/smf-go/src/smf
```

3. Create the following flatbuffer file (`not_a_demo.fbs`):

```
namespace not_a_demo;

table Request1 {
  name: string;
}

table Response1 {
  name: string;
}

rpc_service SmfStorage1 {
  Get(Request1):Response1;
}
```

3. Invoke `flatc` and `smfc` on it (ensure you're in the `${GOPATH}/src` directory):

```
/path/to/flatc --gen-name-strings --gen-object-api --go --force-empty --gen-compare --keep-prefix --json --reflect-names --defaults-json --gen-mutable -o . /path/to/not_a_demo.fbs 

/path/to/smfc --filename not_demo_service.fbs --language=go --output_path=.
```

4. Make a directory for the "main" module:

```
mkdir not_a_demo_client
```

5. Create a `not_a_demo_client/main.go` file containing the following:

```go
package main

import (
        "fmt"
        "not_a_demo"
)

func main() {
        fmt.Printf("It compiled!")
}
```

6. `cd` to `not_a_demo_client` and attempt a build:

```
cd not_a_demo_client
go build
```

The compilation fails as above.

**Expected behavior**
Output code to contain correct descriptions of types, and thusly, for compilation to succeed.

** Environment Info **

- OS: Ubuntu 16.04
- smf version: git-9dfa658
- smf-go version: git-91025c4
- Compiler: g++-7
- Go Version: 1.11.2

--- a/src/smfc/go_generator.cc
+++ b/src/smfc/go_generator.cc
@@ -17,6 +17,7 @@ go_public_name(std::string s) {
   s[0] = std::toupper(s[0]);
   return s;
 }
+
 static void
 print_server_method(smf_printer &printer, const smf_method *method) {
   std::map<std::string, std::string> vars;
@@ -25,14 +26,15 @@ print_server_method(smf_printer &printer, const smf_method *method) {
   vars["MethodID"] = std::to_string(method->method_id());
   vars["RawMethodName"] = "Raw" + go_public_name(method->name());
   vars["MethodName"] = go_public_name(method->name());
+  vars["InType"] = method->input_type_name(language::go);
 
   printer.print(vars, "func (s *$ServiceName$) $RawMethodName$(ctx "
                       "context.Context, req []byte) ([]byte, error) {\n");
   printer.indent();
   printer.print(vars, "return s.$InterfaceName$.$MethodName$(ctx, "
-                      "demo.GetRootAsRequest(req, 0))\n");
+                      "GetRootAs$InType$(req, 0))\n");
   printer.outdent();
-  printer.print("}\n");
+  printer.print("}\n\n");
 }
 
 static void
@@ -62,7 +64,7 @@ print_server_jump_table(smf_printer &printer, const smf_service *service) {
   printer.print("}\n");
 
   printer.outdent();
-  printer.print("}\n");
+  printer.print("}\n\n");
 }
 static void
 print_server(smf_printer &printer, const smf_service *service) {
@@ -72,18 +74,20 @@ print_server(smf_printer &printer, const smf_service *service) {
   vars["InterfaceName"] = go_public_name(service->name());
   vars["ServiceName"] = vars["InterfaceName"] + "Service";
   vars["ServiceID"] = std::to_string(service->service_id());
-  printer.print("\n// Server\n");
+  printer.print(vars,
+                "\n// $InterfaceName$ - $InterfaceName$ Server interface\n");
 
   // gen the interface for service
   printer.print(vars, "type $InterfaceName$ interface {\n");
   printer.indent();
   for (auto &m : service->methods()) {
     vars["MethodName"] = go_public_name(m->name());
-    printer.print(
-      vars, "$MethodName$(context.Context, *smf.Request) ([]byte, error)\n");
+    vars["InType"] = m->input_type_name(language::go);
+    printer.print(vars,
+                  "$MethodName$(context.Context, *$InType$) ([]byte, error)\n");
   }
   printer.outdent();
-  printer.print("}\n");
+  printer.print("}\n\n");
 
   // gen the service
   printer.print(vars, "type $ServiceName$ struct {\n");
@@ -98,21 +102,21 @@ print_server(smf_printer &printer, const smf_service *service) {
   printer.indent();
   printer.print(vars, "return &$ServiceName${$InterfaceName$: s}\n");
   printer.outdent();
-  printer.print("}\n");
+  printer.print("}\n\n");
 
   // gen service name
   printer.print(vars, "func (s *$ServiceName$) ServiceName() string {\n");
   printer.indent();
   printer.print(vars, "return \"$InterfaceName$\"\n");
   printer.outdent();
-  printer.print("}\n");
+  printer.print("}\n\n");
 
   // gen service id
   printer.print(vars, "func (s *$ServiceName$) ServiceID() uint32 {\n");
   printer.indent();
   printer.print(vars, "return $ServiceID$\n");
   printer.outdent();
-  printer.print("}\n");
+  printer.print("}\n\n");
 
   // gen jump table
   print_server_jump_table(printer, service);
@@ -145,9 +149,9 @@ print_client_method(smf_printer &printer, const smf_method *method) {
   printer.print("return nil, err\n");
   printer.outdent();
   printer.print("}\n");
-  printer.print(vars, "return smf.GetRootAsResponse(res, 0), nil\n");
+  printer.print(vars, "return GetRootAs$OutputType$(res, 0), nil\n");
   printer.outdent();
-  printer.print("}\n");
+  printer.print("}\n\n");
 
   // raw
   printer.print(vars, "func (s *$ClientName$) $RawMethodName$(ctx "
@@ -168,7 +172,7 @@ print_client(smf_printer &printer, const smf_service *service) {
   vars["ClientName"] = vars["InterfaceName"] + "Client";
   vars["ServiceID"] = std::to_string(service->service_id());
 
-  printer.print("\n// Client\n");
+  printer.print("// Client\n");
   // gen the client
   printer.print(vars, "type $ClientName$ struct {\n");
   printer.indent();
@@ -181,7 +185,7 @@ print_client(smf_printer &printer, const smf_service *service) {
   printer.indent();
   printer.print(vars, "return &$ClientName${Client: c}\n");
   printer.outdent();
-  printer.print("}\n");
+  printer.print("}\n\n");
 
   // gen all methods
   for (auto &m : service->methods()) {
@@ -202,6 +206,7 @@ go_generator::generate_header_prologue() {
   printer_.print(vars, "// source: $filename$\n");
   printer_.print(vars, "package $package$\n\n");
 }
+
 void
 go_generator::generate_header_includes() {
   VLOG(1) << "get_header_includes";
@@ -213,6 +218,7 @@ go_generator::generate_header_includes() {
   printer_.outdent();
   printer_.print(")\n");
 }
+
 void
 go_generator::generate_header_services() {
   VLOG(1) << "Generating (" << services().size() << ") services";

--- a/src/smfc/go_generator.h
+++ b/src/smfc/go_generator.h
@@ -41,10 +41,6 @@ class go_generator : public generator {
     generate_header_includes();
     generate_header_services();
 
-    // for Go make sure that the directories exist
-    boost::filesystem::create_directories(
-      boost::algorithm::join(package_parts(), "/"));
-
     return save_conents_to_file();
   }
 

--- a/src/smfc/smf_method.h
+++ b/src/smfc/smf_method.h
@@ -45,19 +45,15 @@ class smf_method {
 
   std::string
   type(const flatbuffers::StructDef &sd, language lang) const {
-    std::deque<std::string> tmp(sd.defined_namespace->components.begin(),
-                                sd.defined_namespace->components.end());
-
-    tmp.push_back(sd.name);
     switch (lang) {
-    case language::cpp:
+    case language::cpp: {
+      std::deque<std::string> tmp(sd.defined_namespace->components.begin(),
+                                  sd.defined_namespace->components.end());
+      tmp.push_back(sd.name);
       return boost::algorithm::join(tmp, "::");
+    }
     case language::go:
-      // for go we only want the last package and the name of method
-      for (auto i = 0u; i < tmp.size() - 2; ++i) {
-        tmp.pop_front();
-      }
-      return boost::algorithm::join(tmp, ".");
+      return sd.name;
     default:
       throw std::runtime_error("Unknown language for the method.h");
     }

